DigitalLearningSolutions.Web/Controllers/LearningPortalController/RecommendedLearningController.cs

```diff
@@ -0,0 +1,211 @@
+﻿namespace DigitalLearningSolutions.Web.Controllers.LearningPortalController
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using DigitalLearningSolutions.Data.Helpers;
+    using DigitalLearningSolutions.Data.Models.External.Filtered;
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Web.Helpers;
+    using DigitalLearningSolutions.Web.Helpers.ExternalApis;
+    using DigitalLearningSolutions.Web.ServiceFilter;
+    using DigitalLearningSolutions.Web.ViewModels.LearningPortal.RecommendedLearning;
+    using DigitalLearningSolutions.Web.ViewModels.LearningPortal.SelfAssessments.FilteredMgp;
+    using Microsoft.AspNetCore.Authorization;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Extensions.Configuration;
+
+    [Authorize(Policy = CustomPolicies.UserOnly)]
+    [ServiceFilter(typeof(VerifyDelegateUserCanAccessSelfAssessment))]
+    public class RecommendedLearningController : Controller
+    {
+        private readonly IConfiguration configuration;
+        private readonly IFilteredApiHelperService filteredApiHelperService;
+        private readonly ISelfAssessmentService selfAssessmentService;
+
+        public RecommendedLearningController(
+            IFilteredApiHelperService filteredApiHelperService,
+            ISelfAssessmentService selfAssessmentService,
+            IConfiguration configuration
+        )
+        {
+            this.filteredApiHelperService = filteredApiHelperService;
+            this.selfAssessmentService = selfAssessmentService;
+            this.configuration = configuration;
+        }
+
+        [Route("/LearningPortal/SelfAssessment/{selfAssessmentId:int}/Results")]
+        public async Task<IActionResult> SelfAssessmentResults(int selfAssessmentId)
+        {
+            var candidateId = User.GetCandidateIdKnownNotNull();
+            
+            selfAssessmentService.SetSubmittedDateNow(selfAssessmentId, candidateId);
+            selfAssessmentService.SetUpdatedFlag(selfAssessmentId, candidateId, false);
+
+            if (!configuration.IsSignpostingUsed())
+            {
+                var filteredToken = await GetFilteredToken();
+                var profile = selfAssessmentService.GetFilteredProfileForCandidateById(selfAssessmentId, candidateId);
+                var goals = selfAssessmentService.GetFilteredGoalsForCandidateId(selfAssessmentId, candidateId)
+                    .ToList();
+                var response = await filteredApiHelperService.UpdateProfileAndGoals(filteredToken, profile, goals);
+            }

```

Comment:
I'm a bit confused why this doesn't take the user straight to the Filtered/Results page if Signposting is turned off, rather than going through the new RecommendedLearning endpoint first.
I would have expected this endpoint to direct the user either to Signposting or to Filtered depending on which was enabled.

Reply:
I've tried to reduce the duplicated code between the 2 actions here, although this may have made it more confusing. We are either submitting and displaying, or just displaying depending on whether the self assessment has unprocessed updates.

Previously both actions contained the contents of the RecommendedLearning action (I've renamed this to remove the Filtered from the name) and this SelfAssessmentResults action contained the extra stuff that still remains. Both actions had the same bookmark etc.

We need to preserve the 2 actions logic from the view since we still care about marking the self assessment as submitted only when it can be but we could potentially change the redirect pattern here if there is a better way to do this.

Reply:
@AlexJacksonDS Okay, I think this is generally alright. Couple of things in here that I'm wondering about
* We're setting the url of the RecommendedLearning endpoint as a bookmark on the user's self assessment, and this code change modifies the url (removes "Filtered") - could this cause problems?
* Can the bit inside the if-statement here be pulled out into a method `UpdateFilteredProfileAndGoalsForDelegate(candidateId)`?

Reply:
I'm not 100% sure on the bookmarks, I guess it might. Maybe we need to have an endpoint that still has the old route and sets the bookmark accordingly, then change that to redirect to the new one?

Reply:
Talked to Kevin about this, and we'd have to run a script on prod to update the bookmarks if we changed the url for this endpoint. We agreed it would be better just to keep it as it is, even though "Filtered" is in the current url.

Reply:
(The bookmark is the url the user gets taken to when they launch the self-assessment.) Given that the endpoint it currently bookmarks updates the submission time, I think it's fine to stick with bookmarking the url of this endpoint. The extra bonus in doing that is that you redirect them to an endpoint with a non-filtered-related url.

Reply:
Does [this logic](https://github.com/TechnologyEnhancedLearning/DLSV2/blob/HEEDLS-698-controller-refactor-temp/DigitalLearningSolutions.Web/Controllers/LearningPortalController/RecommendedLearningController.cs) look any better? I've re-added the Filtered/Dashboard action, which will redirect to the new one when signposting is on.

The /Results endpoint should be fine to change, it was never bookmarked, so we can't break the continue button on the first page of the self assessment which is the only place that the bookmark is used.

Reply:
@AlexJacksonDS Think the pattern looks about right. Thoughts:
* We should move the `if (configuration.IsSignpostingUsed())` to the top of `RecommendedLearning()` to mirror what happens in `FilteredDashboard`. 
    * On a related note, we can just redirect from RecommendedLearning() to FilteredDashboard() in this case.